zerocopy-derive/src/enums.rs

```diff
@@ -0,0 +1,359 @@
+                            }
+                        )
+                    };
+                    // SAFETY: `cast_unsized` removes the initialization
+                    // invariant from `p`, so we re-assert that all of the bytes
+                    // are initialized.
+                    let variant = unsafe { variant.assume_initialized() };
+                    <
+                        #variant_struct_ident #ty_generics as #trait_path
+                    >::is_bit_valid(variant)
+                }
+            }
+        }
+    });
+
+    quote! {
+        // SAFETY: We use `is_bit_valid` to validate that the bit pattern of the
+        // enum's tag corresponds to one of the enum's discriminants. Then, we
+        // check the bit validity of each field of the corresponding variant.
+        // Thus, this is a sound implementation of `is_bit_valid`.
+        fn is_bit_valid<A>(
+            mut candidate: ::zerocopy::Maybe<'_, Self, A>,
+        ) -> ::zerocopy::macro_util::core_reexport::primitive::bool
+        where
+            A: ::zerocopy::pointer::invariant::Aliasing
+                + ::zerocopy::pointer::invariant::AtLeast<::zerocopy::pointer::invariant::Shared>,
+        {
+            use ::zerocopy::macro_util::core_reexport;
+
+            #tag_enum
+
+            type ___ZerocopyTagPrimitive = ::zerocopy::macro_util::SizeToTag<{
+                core_reexport::mem::size_of::<___ZerocopyTag>()
+            }>;
+
+            #tag_consts
+
+            type ___ZerocopyOuterTag = #outer_tag_type;
+            type ___ZerocopyInnerTag = #inner_tag_type;
+
+            #variant_structs
+
+            #variants_union
+
+            #[repr(C)]
+            struct ___ZerocopyRawEnum #generics {
+                tag: ___ZerocopyOuterTag,
+                variants: ___ZerocopyVariants #ty_generics,
+            }
+
+            let candidate = candidate.reborrow().forget_exclusive();
+
+            // SAFETY:
+            // - `cast` is implemented as required.
+            // - By definition, `*mut Self` and `*mut __ZerocopyTagPrimitive`
+            //   are types of the same size.
+            let tag = unsafe {
+                candidate.cast_unsized(|p: *mut Self| {
+                    p as *mut ___ZerocopyTagPrimitive
+                })
+            };
+            // SAFETY: Since `candidate` has the invariant `Initialized`, we
+            // know that `candidate`'s referent (and thus `tag`'s referent) is
+            // as-initialized as `Self`. Since all of the possible tag types
+            // always have all of their bytes initialized, we know that
+            // `tag`'s referent has all of its bytes initialized. Since the
+            // validity invariant of `__ZerocopyTagPrimitive` is just that all
+            // of its bytes are initialized, we know that `tag`'s referent is
+            // bit-valid.
+            let tag = unsafe { tag.assume_valid() };
+            let tag = tag.read_unaligned();
```

Comment:
We can simplify this proof a bit by calling the unsafe `tag.assume_initialized()` — the justification for this is just that `candidate` is `Initialized` and casting it doesn't clear invariant. Then, call `tag.bikeshed_recall_valid()`, which is safe and handles the proof that initialized implies valid for the tag type.

Reply:
I agree, but `bikeshed_recall_valid` is `pub(crate)`.

Reply:
Add `#[inline]`, `#[doc(hidden)]`, and make it `pub`! :-)

Reply:
Sure. I assumed that since it was named `bikeshed` that you weren't interested in making it public yet, so hiding it seems sensible. In the interest of bikeshedding, I'd call this function `always_valid` since it's using the fact that the type is unconditionally valid to add the validity invariant to the pointer.

Reply:
The `bikeshed` is just there for until we remove the `doc(hidden)`. Previous contenders have included `recall_valid` (as a counterpart to `forget_valid`). Josh had another idea that I can't remember.

For now, just keep the `bikeshed`.

Reply:
Yeah, to clarify I have no intention of renaming it in this PR.